File: Java/src/main/java/com/sts/slaythesquire/sockets/Packet.java
package com.sts.slaythesquire.sockets;

import com.google.gson.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Packet {

    private String action;
    private HashMap<String, String> properties = new HashMap<>(); // name, json

    /**
     * <summary>
     *     Use this for new Packets
     * </summary>
     */
    public Packet() {}

    /**
     * <summary>
     *     Use this for recieved Packets
     * </summary>
     * @param json The recieved json
     */
    public Packet(String json) {
        Gson gson = new Gson();
        JsonObject jsonObject = new JsonParser().parse(json.trim()).getAsJsonObject();

        action = gson.fromJson(jsonObject.get("action"), String.class);

        JsonObject props = jsonObject.getAsJsonObject("properties");
        if (props != null) {
            for (Map.Entry<String, JsonElement> entry : props.entrySet()) {
                properties.put(entry.getKey(), entry.getValue().getAsString());
            }
        }
    }

    /**
     *
     * @return Returns the json to send
     */
    public String getMessage() {
        JsonObject props = null;
        if (properties.size() > 0) {
            props = new JsonObject();
            for (Map.Entry<String, String> entry : properties.entrySet()) {
                props.addProperty(entry.getKey(), entry.getValue());
            }
        }

        JsonObject json = new JsonObject();
        json.addProperty("action", action);
        json.add("properties", props);

        return json.toString();
    }

    public void setAction(String action) {
        this.action = action.toUpperCase();
    }

    public String getAction() {
        return this.action;
    }

    public void overrideProperty(String property, String json) {
        removeProperty(property);
        addProperty(property, json);
    }

    public boolean addProperty(String property, String json) {
        if (properties.containsKey(property))
            return false;

        properties.put(property, json);
        return true;
    }

    public boolean addProperty(String property, int value){
        return addProperty(property, Integer.toString(value));
    }

    public boolean addArrayProperty(String property, List<String> jsonList) {
        if (properties.containsKey(property))
            return false;

        JsonArray array = new JsonArray();
        for (String json : jsonList) {
            array.add(json);
        }

        properties.put(property, array.toString());
        return true;
    }

    public boolean removeProperty(String property) {
        if (!properties.containsKey(property))
            return false;

        properties.remove(property);
        return true;
    }

    public String getProperty(String property) {
        if (!properties.containsKey(property))
            return null;

        return properties.get(property);
    }

    public List<String> getArrayProperty(String property) {
        if (!properties.containsKey(property))
            return null;

        List<String> jsonList = new ArrayList<>();

        JsonElement element = new JsonParser().parse(properties.get(property));
        JsonArray array = element.getAsJsonArray();
        for (JsonElement json : array) {
            jsonList.add(json.getAsString());
        }

        return jsonList;
    }

    public HashMap<String, String> getProperties() {
        return properties;
    }

    public void setProperties(HashMap<String, String> properties) {
        this.properties = properties;
    }
}

File: Unity/Assets/Scripts/WebSockets/Packet.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Packet
{

    private string action;
    private Dictionary<string, string> properties;

    public string Action
    {
        get
        {
            return action;
        }
        set
        {
            action = value.ToUpper();
        }
    }
    /// <summary>
    /// Returns the json to send
    /// </summary>
    public string Message
    {
        get
        {
            JObject props = props = new JObject();
            foreach (KeyValuePair<string, string> item in properties)
            {
                props.Add(item.Key, item.Value);
            }

            JObject json = new JObject(
                new JProperty("action", action),
                new JProperty("properties", props));

            return json.ToString();
        }
    }

    /// <summary>
    /// Use this for new Packets
    /// </summary>
    public Packet()
    {
        properties = new Dictionary<string, string>();
    }

    /// <summary>
    /// Use this for recieved Packets
    /// </summary>
    /// <param name="recievedMessage"></param>
    public Packet(string message)
    {
        properties = new Dictionary<string, string>();

        JObject json = JObject.Parse(message);

        action = json.GetValue("action").ToString();

        JToken propToken = json.GetValue("properties");
        if (propToken.HasValues)
        {
            JObject props = (JObject)propToken;
            foreach (var item in props)
            {
                properties.Add(item.Key, item.Value.ToString());
            }
        }
    }

    public void OverrideProperty(string property, string value)
    {
        RemoveProperty(property);
        AddProperty(property, value);
    }

    public void OverrideArrayProperty(string property, List<string> value)
    {
        RemoveProperty(property);
        AddArrayProperty(property, value);
    }

    public bool AddProperty(string property, int value)
    {
        return AddProperty(property, value.ToString());
    }

    public bool AddProperty(string property, string value)
    {
        if (!properties.ContainsKey(property))
        {
            properties.Add(property, value);
            return true;
        }
        return false;
    }

    public bool AddArrayProperty(string property, List<string> value)
    {
        if (!properties.ContainsKey(property))
        {
            string[] valueArray = value.ToArray();
            JArray array = JArray.FromObject(valueArray);
            properties.Add(property, array.ToString());
            return true;
        }
        return false;
    }

    public bool RemoveProperty(string property)
    {
        if (properties.ContainsKey(property))
        {
            properties.Remove(property);
            return true;
        }
        return false;
    }

    public string GetProperty(string property)
    {
        if (properties.ContainsKey(property))
        {
            return properties[property];
        }
        return null;
    }

    public List<string> GetArrayProperty(string property)
    {
        if (properties.ContainsKey(property))
        {
            List<string> list = new List<string>();

            JArray array = JArray.Parse(properties[property]);
            foreach (JToken token in array.Children())
            {
                string json = token.ToString();
                Debug.Log(json);
            }

            return list;
        }
        return null;
    }
}